lib/src/boundimage.rs

```diff
@@ -0,0 +1,171 @@
+use crate::task::Task;
+use anyhow::{Context, Result};
+use camino::Utf8Path;
+use fn_error_context::context;
+use ostree_ext::ostree::Deployment;
+use ostree_ext::sysroot::SysrootLock;
+use regex::Regex;
+use rustix::fs::{OFlags, ResolveFlags};
+use std::fs;
+use std::fs::File;
+use std::io::Read;
+use std::os::unix::io::AsFd;
+use std::path::Path;
+
+const BOUND_IMAGE_DIR: &'static str = "usr/lib/bootc/bound-images.d";
+
+pub(crate) struct BoundImageManager {
+    deployment_root: String, // full path to the deployment root
+    spec_dir: String,        //full path to the bound image spec dir
+}
+
+impl BoundImageManager {
+    pub(crate) fn new(deployment: &Deployment, sysroot: &SysrootLock) -> Result<BoundImageManager> {
+        let deployment_root = sysroot.deployment_dirpath(&deployment);
+
+        let bound_image_manager = BoundImageManager {
+            deployment_root: deployment_root.to_string(),
+            spec_dir: format!("/{}/{BOUND_IMAGE_DIR}", deployment_root),
+        };
+
+        Ok(bound_image_manager)
+    }
+
+    pub(crate) fn run(&self) -> Result<()> {
+        if Path::new(&self.spec_dir).exists() {
+            let bound_images = self.parse_spec_dir()?;
+            self.pull_images(bound_images)?;
+        }
+
+        Ok(())
+    }
+
+    #[context("parse bound image spec dir")]
+    fn parse_spec_dir(&self) -> Result<Vec<BoundImage>> {
+        let entries = fs::read_dir(&self.spec_dir)?;
+        let mut bound_images = Vec::new();
+
+        for entry in entries {
+            //validate entry is a symlink with correct extension
+            let entry = entry?;
+            let file_name = entry.file_name();
+            let file_name = if let Some(n) = file_name.to_str() {
+                n
+            } else {
+                anyhow::bail!(
+                    "Invalid non-UTF8 filename: {file_name:?} in {}",
+                    &self.spec_dir
+                );
+            };
+
+            if !entry.file_type()?.is_symlink() {
+                anyhow::bail!("Not a symlink: {file_name}");
+            }
+
+            //parse the file contents
+            let file_path = entry.path();
+            let file_path = file_path.strip_prefix(format!("/{}", &self.deployment_root))?;
+
+            let root_dir = File::open(&self.deployment_root)?;
+            let root_fd = root_dir.as_fd();
+
+            let fd = rustix::fs::openat2(
+                root_fd,
+                file_path,
+                OFlags::empty(),
+                rustix::fs::Mode::empty(),
+                ResolveFlags::IN_ROOT,
+            )?;
+
+            let mut file = { File::from(fd) };
```

Comment:
Minor style thing, I think this would be clearer as `let fd: File = rustix::fs::openat2()?.into()` alongside the above.
(Dealing with `OwnedFd` vs `File` is a weird part of dropping down to the rustix level)